Document backward incompatible changes from dependencies upgrade and (#129)
front-commerce/developers.front-commerce.com: source/docs/appendices/migration-guides.md

See [this commit](https://gitlab.com/front-commerce/front-commerce/-/commit/bf9f9ace04bc24a7391fe36c267dd94857663db0) and [this one](https://gitlab.com/front-commerce/front-commerce/-/commit/60c31720fe3f113961555c493417b8c8bd45509b) for an overview of those changes.

### Make linting pass again

Dependencies were updated, a first step would be to automatically update as much as your codebase as possible to ensure it matches the latest conventions and good practices.
You can run:

```shell
npm run lint -- --fix
```

Below are the main updates to be aware of.

#### Prettier updated to 2.x

We've updated Prettier to its latest version (1.x -> 2.x). It will not cause any issues to your application, but may break snapshot tests. Your code will have to be reformatted with the latest Prettier version.

It introduced trailing commas, parents for single param arrow functions and other changes. You can read [Prettier 2.0 "2020" blog post](https://prettier.io/blog/2020/03/21/2.0.0.html) for further details.

#### ESLint and rules updated to 7.x

> ESLint (and related rules) have been updated to their latest version. Please ensure your code match the latest rules if your app uses Front-Commerce's default rules. `lint-staged` has also been updated from 9.x to 10.x and requires node 10.13+.

Reviewer: Lint staged was never set up in the skeleton to begin with. Not sure what we should do regarding the documentation in this case. I think I wouldn't mention it.

Author: Sounds good to me, it is also not very important.

See [Migrating to v7.0.0](https://eslint.org/docs/user-guide/migrating-to-7.0.0) for further details.

### ElasticSearch related changes

#### `@elastic/elasticsearch` must be installed

Recent versions of the official [`@elastic/elasticsearch`](https://www.npmjs.com/package/@elastic/elasticsearch) client are versioned against the corresponding ElasticSearch version they are compatible with. We decided to make it a [peer-dependency](https://nodejs.org/es/blog/npm/peer-dependencies/) of `front-commerce`'s package instead of being a direct dependency to allow developers to choose the exact version they need.

You must install it in your application, with the version matching your server version. For now, only 6.x is supported.

```shell
npm install --save @elastic/elasticsearch@6
```

> **This library had several breaking changes.** The most important one is that `client.search` now returns an object instead of the body.

Reviewer: I'd add a reassuring message telling the users that this will affect them only if they changed how products were retrieved from ES. If they only use configurations available in `website.js` they don't have to worry about this update.

### GraphQL Dataloaders: `loadMany()` change

The [`dataloader`](https://www.npmjs.com/package/dataloader) library has been updated to 2.0.0. There is only one important API change to be aware of.

The `loadMany()` method will now return an array that may contain Error instances along with valid results when a key failed instead of rejecting the promise. Please read [the release notes](https://github.com/graphql/dataloader/releases/tag/v2.0.0) for further information.

For a fully backward compatible change, you can refactor your code as below (or leverage this new feature for a more resilient app):

```diff
const ids = [1, 2, 3];

- return loader.loadMany(ids);
+ return Promise.all(ids.map(loader.load));
```

### `graphql-tools` now in 6.x: remote schema stitching need changes

`graphql-tools` has been migrated to its latest version (v6). This is a huge improvement for the library: after years of slow developments, the project has been taken over from Apollo by the Guild team. Front-Commerce now uses modules from the `@graphql-tools` monorepo, a new and fresh version of this library. See https://the-guild.dev/blog/graphql-tools-v6 for more context.

> **You shouldn't worry if using Front-Commerce's GraphQL modules.** However, if using remote schema stitching features, you will have to update your code according to our latest documentation: see [GraphQL Remote schemas](https://developers.front-commerce.com/docs/advanced/graphql/remote-schemas.html) for an updated guide.

Reviewer: ```suggestion
**You shouldn't worry if you are using Front-Commerce's GraphQL modules.** However, if you are using remote schema stitching features, you will have to update your code according to our latest documentation: see [GraphQL Remote schemas](https://developers.front-commerce.com/docs/advanced/graphql/remote-schemas.html) for an updated guide.
```

Here is an example from [Front-Commerce's Magento 2 GraphQL remote schema module's migration](https://gitlab.com/front-commerce/front-commerce/-/commit/964cb0dee8b614053c9147997f26c07327e61a0a#e9c8fef58455edd870533da439f028dca7618e8c_1_1):

```diff
- import { FilterRootFields } from "graphql-tools";
- import authenticateRequest from "./authenticateRequest";
+ import { FilterRootFields } from "@graphql-tools/wrap";
+ import makeExecutor from "server/core/graphql/makeExecutor";
+ import withMagentoAuthorizationHeaders from "./withMagentoAuthorizationHeaders";
import log from "../../../../scripts/log";

// […]

    const uri = config.magento.endpoint + "/graphql";

    return {
      uri: uri,
      transforms: [
        new FilterRootFields(
          (operation, rootField) =>
            operation === "Query" && rootField === "storeConfig"
        ),
      ],
-      linkContextBuilders: [authenticateRequest()],
+      executor: makeExecutor(uri, {
+        fetchOptionsAdapter: withMagentoAuthorizationHeaders(),
+      }),
    };
  },
};
```

### `axios`: double-check custom interceptors

> `axios` has been updated from 0.15.3 to 0.19.2.

Reviewer: ```suggestion
`axios` has been updated from 0.15.3 to 0.19.2. This change won't affect you if you are using the default Front-Commerce factories when using axios. If you did, here is what will impact you:
```

Another low-level change from 0.17 is that the base url is now prepended to the `config.url` [**AFTER** interceptors (and not *BEFORE*)](https://github.com/axios/axios/pull/950/files#diff-91dcec0516f33811ee5fa71297160b3bL40). **If your app relies on custom interceptors, please ensure they are still working correctly.**
You can use the `finalUrlFromConfig` helper function from a helper module added to mimic `axios` core feature. See [this commit](https://gitlab.com/front-commerce/front-commerce/-/commit/3fff1a398f7b33d56546b987e15c5f5dc6d43524#24dda67967f430db4a7fe8010764e27d5d7d01b6_63_67) for an example.

### `react-intl`: changes for strings containing HTML tags

`react-intl` was updated from 3.x to 4.x. The main breaking change is that missing tags will now throw errors if your i18n messages contain HTML tags. Please [pass additional rendering parameters for these tags](https://formatjs.io/docs/react-intl/upgrade-guide-4x#migrating-off-embedded-html-in-messages) in your `intl.formatMessage` calls.

```diff
- intl.formatMessage("Hello <strong>Bob</strong>");
+ intl.formatMessage("Hello <strong>Bob</strong>", {

> +   strong: (text) => <strong>{text}</strong>,

Reviewer: IIRC the only place where we've used this feature in the base theme is for the Cart Recap. It might be nice tell the users to look for them? I feel like this part does not explain what is expected from our users.

I'd say:
* they should check your translation files and look for html tags
*  they should get the ids of the relevant translations and check that the html tags is correctly referenced where these ids are used.
* we provide the list of the ids that by default have an html tag in front-commerce and tell the users to ensure they don't have any files using these ids or change these files if needed

(my sentences are pretty bad but that's the idea of it)

The react library has bigger changes. Front-Commerce is now using the official and rebranded [@stripe/react-stripe-js](https://www.npmjs.com/package/@stripe/react-stripe-js) instead of the deprecated [react-stripe-elements](https://github.com/stripe/react-stripe-elements) library. If you overrode Stripe components, please check [the "migrating from react-stripe-elements" guide](https://github.com/stripe/react-stripe-js/blob/master/docs/migrating.md) for migration instructions (imports etc…).

We rewrote default components to use new Stripe components:
- the `<StripeCheckoutElement>` has been simplified in FC, please reapply your changes to the latest version if you overrode it
- the `<StripeCheckout>` component was totally rewritten to use hooks. It is very unlikely that you overrode it, but if you did please update your override.

### Miscellaneous and unlikely impacting backwards incompatible changes

Some minor changes were introduced in dependencies or while removing deprecated code. They are unlikely impacting a standard application, but worth mentioning in case you relied on some low-level APIs.

> - The `<ExistingAddress />` and `<NewAddress />` components used in the checkout were updated to use the `<Checkbox />` component as they were using a deprecated one. Make sure it doesn't have an impact on your styles.

Reviewer: This also means that any one that overrode these two components will have a broken build because the component no longer exists. That's not unlikely to happen imo.

Author: Yes, not very likely but I'll add a few words to make it explicit.

- `graphql` has been upgraded from 14.6.0 to 15.1.0. It is very likely backwards compatible for your app, even though backward incompatible changes were introduced for subtle use cases: see [v15.0.0 release notes](https://github.com/graphql/graphql-js/releases/tag/v15.0.0) if curious
- `SitemapLoader` (and `makeMagentoPaginationWalker` helper) were removed from `magento1` and `magento2` modules (file `magento(1|2)/store/loaders`). They were unused in the core. If your application used them, please refactor it to use the loader from Front-Commerce's core. See [our documentation about Sitemap](https://developers.front-commerce.com/docs/advanced/production-ready/sitemap.html#Add-dynamic-pages) to learn about the feature.
- The `ProductStockLoader` no longer takes the `FeatureFlag` loader as parameter. Please remove this parameter if you were instantiating it manually.
- `loaders.Url.matchBy` has been removed from Magento2's module. It was only used for the deprecated `Query.matchUrls` field and might not impact your codebase, but **it's worth mentioning in case your relied on it since it wasn't issuing deprecation warnings itself.**

> - Convict has been upgraded to 6.0.0. You may have to install additional packages if your application uses custom configurations with one of the following format: `"email"`, `"ipaddress"`, `"url"`, `"duration"` or `"timestamp"`. See [their migration documentation](https://github.com/mozilla/node-convict/blob/master/packages/convict/MIGRATING_FROM_CONVICT_5_TO_6.md) for further information

Reviewer: ```suggestion
- Convict has been upgraded to 6.0.0. You may have to install additional packages if your application uses custom configuration providers with one of the following format in their schema: `"email"`, `"ipaddress"`, `"url"`, `"duration"` or `"timestamp"`. See [their migration documentation](https://github.com/mozilla/node-convict/blob/master/packages/convict/MIGRATING_FROM_CONVICT_5_TO_6.md) for further information
```

People using our configuration providers don't really know it's convict under the hood. By adding these two notions it has more changes to ring a bell.
